server/schedule/operator/create_operator.go

```diff
@@ -62,10 +63,23 @@ func CreateForceTransferLeaderOperator(desc string, cluster opt.Cluster, region
 }
 
 // CreateMoveRegionOperator creates an operator that moves a region to specified stores.
-func CreateMoveRegionOperator(desc string, cluster opt.Cluster, region *core.RegionInfo, kind OpKind, peers map[uint64]*metapb.Peer) (*Operator, error) {
-	return NewBuilder(desc, cluster, region).
-		SetPeers(peers).
-		Build(kind)
+func CreateMoveRegionOperator(desc string, cluster opt.Cluster, region *core.RegionInfo, kind OpKind, peers map[uint64]*metapb.Peer, roles map[uint64]placement.PeerRoleType) (*Operator, error) {
+	builder := NewBuilder(desc, cluster, region).SetPeers(peers)
+	for storeID, role := range roles {
+		switch role {
+		case placement.Leader:
+			builder.SetLeader(storeID)
+		case placement.Follower:
+			// if all roles are follower, the originLeaderStoreID will be selected in prepare build phase
+			if builder.targetLeaderStoreID == storeID {
+				builder.targetLeaderStoreID = 0
+			}
```

Comment:
Maybe this is dead code? The newly created builder has `targetLeaderStoreID = 0` by default, and only `setLeader` can modify it in this function. I wonder whether this can stop the builder from choosing the store as a leader.

For example, suppose you have `1(follower), 3(follower), 4(leader)`, then `transfer-region <region> 1 voter 2 voter 4 follower`, you will get `1(follower), 2(follower), 4(leader)`, which is not expected.

Reply:
Right, setting `targetLeaderStoreID` here is redundant.

Reply:
> I suppose a better plan in the given case is 1(leader), 3(follower), 4(follower) since store 1 already has a peer belonging to the region?

Sounds good. IMO, the point is that as long as you specify store 4 as follower, it should never be assigned as leader.

You may want to take a look at `server/schedule/operator/builder`:`buildStepsWith(out)JointConsensus` to enforce the roles.